### PR TITLE
Adding version sorting to display the version completion list correct

### PR DESCRIPTION
**Pull request by YangLyu89**
Libman don't have a reliable version sorting for library versions completion. This can also resolve https://github.com/aspnet/LibraryManager/issues/151

### CODE REVIEW

File: src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs [hunk 1]
```diff
@@ -84,7 +84,7 @@ public CdnjsCatalog(CdnjsProvider provider)
                     {
                         var completion = new CompletionItem
                         {
-                            DisplayText = version,
+                            DisplayText = $"{name}@{version}",
```
Comment (YangLyu89) [inline src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs]: >DisplayText = $"{name}@{version}", [](start = 28, length = 34)

Thinking of make the completion list for cdnjs and unpkg to be consistent.
The name@version will display the whole completion string while version only display the version part.

File: src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj [hunk 2]
```diff
@@ -59,6 +59,7 @@
   </PropertyGroup>
   <ItemGroup>
     <Compile Include="Commands\InstallLibraryCommand.cs" />
+    <Compile Include="Json\Completion\LibraryIdCompletionEntry.cs" />
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj]: ><Compile Include="Json\Completion\LibraryIdCompletionEntry.cs" /> [](start = 4, length = 65)

Place next to others in same directory

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        private void SetVersion(string insertionText)
+        {
+            if (!string.IsNullOrEmpty(insertionText))
+            {
+                int atIndex = insertionText.IndexOf('@');
+
+                Debug.Assert(atIndex >= 0);
+
+                string versionText = insertionText.Substring(atIndex + 1);
+                string[] versionParts = versionText.Split('.');
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: >string[] versionParts = versionText.Split('.'); [](start = 16, length = 47)

Would the System.Version class work for this instead of parsing out the parts yourself?

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        protected override int InternalCompareTo(CompletionEntry other)
+        {
+            if (other == null)
+            {
+                return 1;
+            }
+
+            LibraryIdCompletionEntry otherEntry = other as LibraryIdCompletionEntry;
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: >otherEntry [](start = 37, length = 10)

You probably want to null check this.

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs [hunk 5]
```diff
@@ -77,17 +76,26 @@ protected override IEnumerable<JSONCompletionEntry> GetEntries(JSONCompletionCon
...
+                    bool versionSpecific = false;
+                    bool setVersionSpecific = true;
+
                     foreach (CompletionItem item in set.Completions)
                     {
+                        if (setVersionSpecific)
+                        {
+                            versionSpecific = item.DisplayText.Contains("@");
+                            setVersionSpecific = false;
+                        }
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: Could the LibraryIdCompletionEntry class do this for you?

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs [hunk 6]
```diff
@@ -77,17 +76,26 @@ protected override IEnumerable<JSONCompletionEntry> GetEntries(JSONCompletionCon
...
+
                         string insertionText = item.InsertionText.Replace("\\\\", "\\").Replace("\\", "\\\\");
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: >string insertionText = item.InsertionText.Replace("\\\\", "\\").Replace("\\", "\\\\"); [](start = 24, length = 86)

This is confusing to me. Can you comment why this is happening?

File: test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs [hunk 7]
```diff
@@ -120,7 +120,7 @@ public void Setup()
             Assert.AreEqual(7, result.Start);
             Assert.AreEqual(0, result.Length);
             Assert.IsTrue(result.Completions.Count() >= 69);
-            Assert.AreEqual("1.2.3", result.Completions.Last().DisplayText);
+            Assert.AreEqual("jquery@1.2.3", result.Completions.Last().DisplayText);
```
Comment (ToddGrun) [inline test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs]: >jquery@1.2.3 [](start = 29, length = 12)

I don't know how this works. Is this information being obtained from a Mock?

File: src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj (hunk 2 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Microsoft.Web.LibraryManager.Vsix.csproj]: Yes. will do

---
In reply to: [198546291](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198546291) [](ancestors = 198546291)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs (hunk 4 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: Yes. I think it is better to null check after that.

---
In reply to: [198551833](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198551833) [](ancestors = 198551833)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 5 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: Yes. This can be moved just using DisplayText

---
In reply to: [198552083](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198552083) [](ancestors = 198552083)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs (hunk 3 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: Looks like I can use the SemanticVersion class.

---
In reply to: [198550380](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198550380) [](ancestors = 198550380)

File: test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs (hunk 7 above)
Comment (YangLyu89) [reply test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs]: I think this is from _catalog.GetLibraryCompletionSetAsync.

---
In reply to: [198553292](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198553292) [](ancestors = 198553292)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 6 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: I don't understand this line. It also confuses me...

---
In reply to: [198552566](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198552566) [](ancestors = 198552566)

File: src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs (hunk 1 above)
Comment (jimmylewis) [reply src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs]: Let's make unpkg match the old behavior here.  Only show either name or version (depending on context), and use name@version for the insertion text in both cases.

---
In reply to: [198361271](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198361271) [](ancestors = 198361271)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.Web.Editor.Completion;
+using Microsoft.VisualStudio.Text;
+using System.Diagnostics;
+using Microsoft.Web.LibraryManager.Providers.Unpkg;
+
+namespace Microsoft.Web.LibraryManager.Vsix
+{
+    internal class LibraryIdCompletionEntry : SimpleCompletionEntry
```
Comment (jimmylewis) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: >LibraryId [](start = 19, length = 9)

I think this should be a VersionCompletionEntry, not LibraryId

**Comment by jimmylewis** [comment]
>                     foreach (CompletionItem item in set.Completions)

We should have a way to determine whether this is name completion or version completion.  Probably the CompletionSet should expose a context.

---
Refers to: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs:79 in 40a016e. [](commit_id = 40a016e256f38766c22848c9a1eb1333d0a70747, deletion_comment = False)

File: src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs [hunk 9]
```diff
@@ -4,16 +4,13 @@
 using Microsoft.JSON.Editor.Completion;
 using Microsoft.VisualStudio.Imaging.Interop;
 using Microsoft.VisualStudio.Language.Intellisense;
-using Microsoft.Web.Editor.Completion;
 using System.Windows.Media;
 using Microsoft.VisualStudio.Text;
 
 namespace Microsoft.Web.LibraryManager.Vsix
 {
     internal class SimpleCompletionEntry : JSONCompletionEntry
```
Comment (jimmylewis) [inline src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs]: >SimpleCompletionEntry [](start = 19, length = 21)

This class should remain as it was in order to let providers customize the order of completion items.

File: src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs (hunk 1 above)
Comment (YangLyu89) [reply src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs]: I have opened an issue to track it.

---
In reply to: [198665798](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198665798) [](ancestors = 198665798,198361271)

**Comment by YangLyu89** [comment]
>                     foreach (CompletionItem item in set.Completions)

Add an enum type field to CompletionItem

---
In reply to: [400855833](https://github.com/aspnet/LibraryManager/pull/258#issuecomment-400855833) [](ancestors = 400855833)

---
Refers to: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs:79 in 40a016e. [](commit_id = 40a016e256f38766c22848c9a1eb1333d0a70747, deletion_comment = False)

File: src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs (hunk 9 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs]: Will do.

---
In reply to: [198666680](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198666680) [](ancestors = 198666680)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs (hunk 8 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionEntry.cs]: As discussed, it will only handle the version entry.

---
In reply to: [198665923](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198665923) [](ancestors = 198665923)

File: test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs (hunk 7 above)
Comment (YangLyu89) [reply test/LibraryManager.Test/Providers/Cdnjs/CdnjsCatalogTest.cs]: Will change it back as to keep the cdnjs display behavior

---
In reply to: [198653075](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198653075) [](ancestors = 198653075,198553292)

**Comment by jimmylewis** [comment]
>                     foreach (CompletionItem item in set.Completions)

I think it might work better on the CompletionSet rather than on each CompletionItem

---
In reply to: [400876094](https://github.com/aspnet/LibraryManager/pull/258#issuecomment-400876094) [](ancestors = 400876094,400855833)

---
Refers to: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs:79 in 40a016e. [](commit_id = 40a016e256f38766c22848c9a1eb1333d0a70747, deletion_comment = False)

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+            if (otherEntry == null)
+            {
+                return 1;
+            }
+
+            if (!string.IsNullOrEmpty(VersionText))
+            {
+                string otherVersionText = otherEntry.VersionText;
+                SemanticVersion selfSemanticVersion = SemanticVersion.Parse(VersionText);
```
Comment (YangLyu89) [inline src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: >SemanticVersion selfSemanticVersion = SemanticVersion.Parse(VersionText); [](start = 16, length = 73)

From SemanticVersion.parse function, it looks like a null check for the versionText will be Okay. Maybe not need a tryParse function.

File: src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs [hunk 11]
```diff
@@ -178,45 +178,51 @@ private void GetFiles(JObject fileObject, List<string> files)
...
-                            string itemText = name + "@" + semVersion.ToString();
-                            CompletionItem completionItem = new CompletionItem
-                            {
-                                DisplayText = itemText,
-                                InsertionText = itemText
-                            };
-
-                            completions.Add(completionItem);
-                        }
+                            DisplayText = semVersion.ToString(),
```
Comment (ToddGrun) [inline src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs]: >semVersion.ToString() [](start = 42, length = 21)

nit: no need to do this twice

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs (hunk 10 above)
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: >SemanticVersion.Parse(VersionText) [](start = 54, length = 34)

Is InternalCompareTo called often? Is SemanticVersion.Parse at all expensive? If the answer to either of these is "yes", then consider moving this to a member variable.

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.Web.Editor.Completion;
+using Microsoft.VisualStudio.Text;
+using System.Diagnostics;
```
Comment (ToddGrun) [inline src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: nit: sort

**Comment by YangLyu89** [comment]
>                     foreach (CompletionItem item in set.Completions)

Adding a boolean before the loop

---
In reply to: [400879889](https://github.com/aspnet/LibraryManager/pull/258#issuecomment-400879889) [](ancestors = 400879889,400876094,400855833)

---
Refers to: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs:79 in 40a016e. [](commit_id = 40a016e256f38766c22848c9a1eb1333d0a70747, deletion_comment = False)

File: src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs (hunk 11 above)
Comment (YangLyu89) [reply src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs]: I see. will change.

---
In reply to: [198903355](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198903355) [](ancestors = 198903355)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs [hunk 13]
```diff
@@ -81,7 +81,15 @@ protected override IEnumerable<JSONCompletionEntry> GetEntries(JSONCompletionCon
                     {
                         string insertionText = item.InsertionText.Replace("\\\\", "\\").Replace("\\", "\\\\");
                         ImageMoniker moniker = item.DisplayText.EndsWith("/") || item.DisplayText.EndsWith("\\") ? _folderIcon : _libraryIcon;
-                        yield return new SimpleCompletionEntry(item.DisplayText, insertionText, item.Description, moniker, trackingSpan, context.Session, ++count);
+
+                        if (set.Type == CompletionType.Version)
+                        {
+                            yield return new VersionCompletionEntry(item.DisplayText, insertionText, item.Description, moniker, trackingSpan, context.Session, ++count);
```
[the remark this answers is not in the document]
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: This is used in SimpleCompletionEntry for sorting the entries. We want to keep the order of the completion entries as the provider sent to us.

---
In reply to: [198904839](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198904839) [](ancestors = 198904839)

[the remark this answers is not in the document]
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: I add it to the VersionCompletionEntry here in case of the VersionCompletionEntry doesn't do the comparison (like empty version text VersionCompletionEntry), so that they can still be handled in SimpleCompletionEntry.

---
In reply to: [198933534](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198933534) [](ancestors = 198933534,198904839)

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs (hunk 12 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: removed this line as it's unused now.

---
In reply to: [198905577](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198905577) [](ancestors = 198905577)

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs (hunk 10 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: I think it's called often. I would obtain this value in the construct function.

---
In reply to: [198905536](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198905536) [](ancestors = 198905536)

**Comment by jimmylewis** [comment]
>             var span = new CompletionSet

rename this to match the type.

---
Refers to: src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs:43 in c416cb5. [](commit_id = c416cb5eed77a562b161e4589d146b4e9ed02ed6, deletion_comment = False)

File: src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs [hunk 14]
```diff
@@ -178,45 +178,52 @@ private void GetFiles(JObject fileObject, List<string> files)
 
             (string name, string version) = LibraryNamingScheme.Instance.GetLibraryNameAndVersion(libraryNameStart);
 
+            int at = name.IndexOf('@');
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs]: Can you add a TODO here to track that the LibraryNamingScheme should not be returning "{name}@" as the name?

File: src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs [hunk 15]
```diff
@@ -178,45 +178,52 @@ private void GetFiles(JObject fileObject, List<string> files)
...
+            name = at > -1 ? name.Substring(0, at) : name;
```
Comment (jimmylewis) [inline src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs]: This should also handle having an @ at the beginning of e.g. a scoped package name.

File: src/LibraryManager.Contracts/CompletionSet.cs [hunk 16]
```diff
@@ -28,6 +28,11 @@ public struct CompletionSet
         /// The list of completions for the span
         /// </summary>
         public IEnumerable<CompletionItem> Completions;
+
+        /// <summary>
+        /// The type of the completion item.
+        /// </summary>
+        public CompletionType Type;
```
Comment (jimmylewis) [inline src/LibraryManager.Contracts/CompletionSet.cs]: name this CompletionType

File: src/LibraryManager.Contracts/CompletionSet.cs [hunk 17]
```diff
@@ -53,4 +58,17 @@ public struct CompletionItem
         /// </summary>
         public string Description;
     }
+
+    /// <summary>
+    /// The completion type is to indicate the type of the completion item.
+    /// </summary>
+    public enum CompletionType
```
Comment (jimmylewis) [inline src/LibraryManager.Contracts/CompletionSet.cs]: >CompletionType [](start = 16, length = 14)

```
public enum CompletionSortOrder
{
    Alphabetical,
    Version,
    AsSpecified,
}
```

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 6 above)
Comment (jimmylewis) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: Can you add a comment:
"Make sure to double-escape any singly-escaped backslashes"

---
In reply to: [198653332](https://github.com/aspnet/LibraryManager/pull/258#discussion_r198653332) [](ancestors = 198653332,198552566)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs [hunk 18]
```diff
@@ -105,7 +115,15 @@ protected override IEnumerable<JSONCompletionEntry> GetEntries(JSONCompletionCon
                             {
                                 string insertionText = item.InsertionText.Replace("\\", "\\\\");
```
Comment (jimmylewis) [inline src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: >Replace [](start = 74, length = 7)

Does this need to do the same .Replace().Replace() as above?

File: src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs [hunk 19]
```diff
@@ -4,9 +4,9 @@
 using Microsoft.JSON.Editor.Completion;
 using Microsoft.VisualStudio.Imaging.Interop;
 using Microsoft.VisualStudio.Language.Intellisense;
-using Microsoft.Web.Editor.Completion;
```
Comment (jimmylewis) [inline src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs]: either revert this change or sort all of these

File: src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+        public VersionCompletionEntry(string displayText, string insertionText, string description, ImageMoniker moniker, ITrackingSpan span, IIntellisenseSession session, int specificVersion)
+            : base(displayText, insertionText, description, moniker, span, session, specificVersion)
+        {
+            if (!string.IsNullOrEmpty(displayText))
+            {
+                SemVersion = SemanticVersion.Parse(displayText);
+            }
+        }
+
+        protected override int InternalCompareTo(CompletionEntry other)
```
Comment (jimmylewis) [inline src/LibraryManager.Vsix/Json/Completion/VersionCompletionEntry.cs]: >InternalCompareTo [](start = 31, length = 17)

Please add unit tests for this.

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 6 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: Sure will do

---
In reply to: [199295753](https://github.com/aspnet/LibraryManager/pull/258#discussion_r199295753) [](ancestors = 199295753,198653332,198552566)

File: src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs (hunk 15 above)
Comment (YangLyu89) [reply src/LibraryManager/Providers/Unpkg/UnpkgCatalog.cs]: Start searching at the second character.

---
In reply to: [199293261](https://github.com/aspnet/LibraryManager/pull/258#discussion_r199293261) [](ancestors = 199293261)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 18 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: changed here.

---
In reply to: [199295856](https://github.com/aspnet/LibraryManager/pull/258#discussion_r199295856) [](ancestors = 199295856)

File: src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs (hunk 19 above)
Comment (YangLyu89) [reply src/LibraryManager.Vsix/Json/Completion/SimpleCompletionEntry.cs]: Sorted.

---
In reply to: [199296051](https://github.com/aspnet/LibraryManager/pull/258#discussion_r199296051) [](ancestors = 199296051)

**Comment by YangLyu89** [comment]
>             var span = new CompletionSet

name as completionSet

---
In reply to: [401486557](https://github.com/aspnet/LibraryManager/pull/258#issuecomment-401486557) [](ancestors = 401486557)

---
Refers to: src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs:43 in c416cb5. [](commit_id = c416cb5eed77a562b161e4589d146b4e9ed02ed6, deletion_comment = False)

File: src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs (hunk 6 above)
Comment (ToddGrun) [reply src/LibraryManager.Vsix/Json/Completion/LibraryIdCompletionProvider.cs]: This is marked as resolved, but it hasn't been explained yet. Can you please clarify in the code why this is happening?

---
In reply to: [199372324](https://github.com/aspnet/LibraryManager/pull/258#discussion_r199372324) [](ancestors = 199372324,199295753,198653332,198552566)

**Comment by YangLyu89** [comment]
I create new PR against rebase issues
